runtime/vm/jnicsup.cpp

```diff
@@ -468,14 +464,27 @@ UDATA JNICALL   pushArguments(J9VMThread *vmThread, J9Method* method, void *args
 		switch (*sigChar++) {
 			case '[':
 				/* skip the rest of the signature */
-				while (*sigChar == '[') sigChar++;
-				if (*sigChar++ != 'L') goto dontSkipClassName;
+				while (*sigChar == '[') {
+					sigChar++;
+				}
+				if ('L' == *sigChar++) {
+					/* skip the rest of the signature */
+					while (';' != *sigChar) {
+						sigChar += 1;					
+					}
+				}
+				sp -= 1;
+				objArg = ARG(jobject, l);
+				*sp = (0 == objArg) ? 0: (UDATA) *((j9object_t*) objArg);
+				break;
```

Comment:
I actually prefer the goto in this case (with an additional comment about intentional fall-through when the goto isn't taken).